File: backend/app/schemas/event.py
"""
Event schemas with timezone-aware request/response.
All stored times are UTC; API can accept local times on create and return local times on read.
"""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.core.utils.timezone import DEFAULT_TZ, normalize_timezone, to_local, to_utc


# ---------- Create (client sends local time + timezone) ----------


class EventCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    start_local: datetime = Field(..., description="Start in event's local timezone (naive or aware)")
    end_local: datetime = Field(..., description="End in event's local timezone")
    timezone: str = Field(default=DEFAULT_TZ, max_length=64)
    lat: float | None = None
    lng: float | None = None
    address: str | None = None
    city: str | None = None
    country_code: str | None = Field(None, max_length=2)
    is_virtual: bool = False
    category: str | None = Field(None, max_length=64)
    image_url: str | None = Field(None, max_length=512)
    max_attendees: int | None = Field(None, ge=0)

    def to_utc_times(self) -> tuple[datetime, datetime]:
        """Return (start_utc, end_utc) for storage."""
        tz = normalize_timezone(self.timezone)
        return to_utc(self.start_local, tz), to_utc(self.end_local, tz)


# ---------- Update (partial) ----------


class EventUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    start_local: datetime | None = None
    end_local: datetime | None = None
    timezone: str | None = None
    lat: float | None = None
    lng: float | None = None
    address: str | None = None
    city: str | None = None
    country_code: str | None = None
    is_virtual: bool | None = None
    category: str | None = None
    image_url: str | None = None
    max_attendees: int | None = None


# ---------- Response (optionally in user's timezone) ----------


class EventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str | None
    start_utc: datetime
    end_utc: datetime
    start_local: datetime | None = None  # set when ?tz= or X-Timezone provided
    end_local: datetime | None = None
    timezone: str
    lat: float | None
    lng: float | None
    address: str | None
    city: str | None
    country_code: str | None
    is_virtual: bool
    category: str | None
    image_url: str | None
    max_attendees: int | None
    is_approved: bool
    created_by: UUID
    created_at: datetime
    views_count: int
    rsvp_count: int

    @classmethod
    def from_orm_with_tz(cls, obj: object, tz: str | None) -> "EventResponse":
        """Build response with start_local/end_local in given timezone."""
        from app.core.utils.timezone import normalize_timezone

        data = cls.model_validate(obj)
        if tz:
            tz = normalize_timezone(tz)
            data.start_local = to_local(data.start_utc, tz)
            data.end_local = to_local(data.end_utc, tz)
        return data


# ---------- List item (summary) ----------


class EventListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    start_utc: datetime
    end_utc: datetime
    timezone: str
    lat: float | None
    lng: float | None
    address: str | None
    city: str | None
    country_code: str | None
    is_virtual: bool
    category: str | None
    image_url: str | None
    rsvp_count: int
    views_count: int

File: frontend/lib/models/event.dart
/// Event model matching backend API (UTC times + optional local display).
class Event {
  final String id;
  final String title;
  final String? description;
  final DateTime startUtc;
  final DateTime endUtc;
  final DateTime? startLocal;
  final DateTime? endLocal;
  final String timezone;
  final double? lat;
  final double? lng;
  final String? address;
  final String? city;
  final String? countryCode;
  final bool isVirtual;
  final String? category;
  final String? imageUrl;
  final int? maxAttendees;
  final bool isApproved;
  final String createdBy;
  final DateTime createdAt;
  final int viewsCount;
  final int rsvpCount;

  const Event({
    required this.id,
    required this.title,
    this.description,
    required this.startUtc,
    required this.endUtc,
    this.startLocal,
    this.endLocal,
    required this.timezone,
    this.lat,
    this.lng,
    this.address,
    this.city,
    this.countryCode,
    required this.isVirtual,
    this.category,
    this.imageUrl,
    this.maxAttendees,
    required this.isApproved,
    required this.createdBy,
    required this.createdAt,
    required this.viewsCount,
    required this.rsvpCount,
  });

  /// Display time: prefer [startLocal] if present (from ?tz= or X-Timezone), else [startUtc] in local.
  DateTime get displayStart => startLocal ?? startUtc.toLocal();
  DateTime get displayEnd => endLocal ?? endUtc.toLocal();

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      startUtc: DateTime.parse(json['start_utc'] as String),
      endUtc: DateTime.parse(json['end_utc'] as String),
      startLocal: json['start_local'] != null
          ? DateTime.parse(json['start_local'] as String)
          : null,
      endLocal: json['end_local'] != null
          ? DateTime.parse(json['end_local'] as String)
          : null,
      timezone: json['timezone'] as String? ?? 'UTC',
      lat: (json['lat'] as num?)?.toDouble(),
      lng: (json['lng'] as num?)?.toDouble(),
      address: json['address'] as String?,
      city: json['city'] as String?,
      countryCode: json['country_code'] as String?,
      isVirtual: json['is_virtual'] as bool? ?? false,
      category: json['category'] as String?,
      imageUrl: json['image_url'] as String?,
      maxAttendees: json['max_attendees'] as int?,
      isApproved: json['is_approved'] as bool? ?? true,
      createdBy: json['created_by'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      viewsCount: json['views_count'] as int? ?? 0,
      rsvpCount: json['rsvp_count'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'start_utc': startUtc.toUtc().toIso8601String(),
      'end_utc': endUtc.toUtc().toIso8601String(),
      'start_local': startLocal?.toIso8601String(),
      'end_local': endLocal?.toIso8601String(),
      'timezone': timezone,
      'lat': lat,
      'lng': lng,
      'address': address,
      'city': city,
      'country_code': countryCode,
      'is_virtual': isVirtual,
      'category': category,
      'image_url': imageUrl,
      'max_attendees': maxAttendees,
      'is_approved': isApproved,
      'created_by': createdBy,
      'created_at': createdAt.toUtc().toIso8601String(),
      'views_count': viewsCount,
      'rsvp_count': rsvpCount,
    };
  }
}

/// List item (summary) from trending / search / nearby.
class EventListItem {
  final String id;
  final String title;
  final DateTime startUtc;
  final DateTime endUtc;
  final String timezone;
  final double? lat;
  final double? lng;
  final String? address;
  final String? city;
  final String? countryCode;
  final bool isVirtual;
  final String? category;
  final String? imageUrl;
  final int rsvpCount;
  final int viewsCount;

  const EventListItem({
    required this.id,
    required this.title,
    required this.startUtc,
    required this.endUtc,
    required this.timezone,
    this.lat,
    this.lng,
    this.address,
    this.city,
    this.countryCode,
    required this.isVirtual,
    this.category,
    this.imageUrl,
    required this.rsvpCount,
    required this.viewsCount,
  });

  factory EventListItem.fromJson(Map<String, dynamic> json) {
    return EventListItem(
      id: json['id'] as String,
      title: json['title'] as String,
      startUtc: DateTime.parse(json['start_utc'] as String),
      endUtc: DateTime.parse(json['end_utc'] as String),
      timezone: json['timezone'] as String? ?? 'UTC',
      lat: (json['lat'] as num?)?.toDouble(),
      lng: (json['lng'] as num?)?.toDouble(),
      address: json['address'] as String?,
      city: json['city'] as String?,
      countryCode: json['country_code'] as String?,
      isVirtual: json['is_virtual'] as bool? ?? false,
      category: json['category'] as String?,
      imageUrl: json['image_url'] as String?,
      rsvpCount: json['rsvp_count'] as int? ?? 0,
      viewsCount: json['views_count'] as int? ?? 0,
    );
  }

  /// For persisting saved events (same shape as API / fromJson).
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'start_utc': startUtc.toUtc().toIso8601String(),
      'end_utc': endUtc.toUtc().toIso8601String(),
      'timezone': timezone,
      'lat': lat,
      'lng': lng,
      'address': address,
      'city': city,
      'country_code': countryCode,
      'is_virtual': isVirtual,
      'category': category,
      'image_url': imageUrl,
      'rsvp_count': rsvpCount,
      'views_count': viewsCount,
    };
  }
}

/// Convert full [Event] to list item for cards and saved list.
EventListItem eventToListItem(Event event) {
  return EventListItem(
    id: event.id,
    title: event.title,
    startUtc: event.startUtc,
    endUtc: event.endUtc,
    timezone: event.timezone,
    lat: event.lat,
    lng: event.lng,
    address: event.address,
    city: event.city,
    countryCode: event.countryCode,
    isVirtual: event.isVirtual,
    category: event.category,
    imageUrl: event.imageUrl,
    rsvpCount: event.rsvpCount,
    viewsCount: event.viewsCount,
  );
}
